src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -739,6 +741,76 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomStringFromDataset : public Generator<std::string> {
+
+public:
+    RandomStringFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()},
+          _pathLength{_path.size()}{}
+
+    void loadDataset() {
+      {
+        if (_pathLength <= 0) {
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+            "ChooseFromDataset requieres non-empty path"));
+            }
+
+        {
+            if (_all_datasets.count(_path) == 0){
+                _ifs.open(_path, std::ifstream::in);
+                if (_ifs.is_open()){
+                    while(std::getline(_ifs, _current_line))
+                    {
+                        // Ignore emtpy lines
+                        if (_current_line.size()!=0){
+                            _all_datasets[_path].push_back(_current_line);
+                        }
+                    }
+                }
+            
+                else {
+                    BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                        "The specified file for ChooseFromDataset cannot be opened or it does not exist")); }
+
+                if (_all_datasets[_path].size() == 0) {
+                    BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                        "The specified file for ChooseFromDataset is empty")); }
+            }
+        }
+      }
+    }
+
+    std::string evaluate() {
+        /* If path the doesn't exist in memory we load it.
+           To avoid reading and storing the datasets multiple times
+           we check if the path already exists in the map (as key). If
+           it doesn't exist, we open the file and store its contents. 
+        */
+
+        if (_all_datasets.count(_path) == 0){
```

Comment:
If two threads fire at the same time they'll both see count=0. Then one will lock, load the dataset, and unlock. Then the other thread will lock, load the dataset (for a second time) and then unlock. I think you want the lock around the size-check:

```c++
{
    std::lock_guard lk{mutex};
    if(sets.count(key) <= 0) {
      load()
    }
}
```

Reply:
I moved the lock just before the check of count, as you explained. Makes more sense, thank you.